src/test/app/Batch_test.cpp

```diff
@@ -0,0 +1,2968 @@
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 0);
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 1),
+                ter(temARRAY_EMPTY));
+            env.close();
+        }
+
+        // temARRAY_TOO_LARGE: Batch: txns array exceeds 8 entries.
+        // telENV_RPC_FAILED
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 9);
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 1),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 2),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 3),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 4),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 5),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 6),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 7),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 8),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 9),
+                ter(telENV_RPC_FAILED));
+            env.close();
+        }
+
+        // temINVALID_INNER_BATCH: Batch: duplicate Txn found.
+        {
+            auto const batchFee = batch::calcBatchFee(env, 1, 2);
+            auto const seq = env.seq(alice);
+            auto jt = env.jtnofill(
+                batch::outer(alice, env.seq(alice), batchFee, tfAllOrNothing),
+                batch::inner(pay(alice, bob, XRP(10)), seq + 1),
+                batch::inner(pay(alice, bob, XRP(10)), seq + 1));
+
+            env(jt.jv, batch::sig(bob), ter(temINVALID_INNER_BATCH));
+            env.close();
+        }
+
+        // temINVALID_INNER_BATCH: Batch: batch cannot have inner batch txn.
+        {
+            auto const seq = env.seq(alice);
+            auto const batchFee = batch::calcBatchFee(env, 0, 2);
+            env(batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner(
+                    batch::outer(alice, seq, batchFee, tfAllOrNothing), seq),
+                batch::inner(pay(alice, bob, XRP(1)), seq + 2),
+                ter(temINVALID_INNER_BATCH));
+            env.close();
+        }
+
+        // temINVALID_INNER_BATCH: Batch: inner txn must have the
+        // tfInnerBatchTxn flag.
+        {
+            auto const batchFee = batch::calcBatchFee(env, 1, 2);
+            auto const seq = env.seq(alice);
+            auto tx1 = pay(alice, bob, XRP(10));
+            tx1[jss::Fee] = "0";
+            tx1[jss::Sequence] = seq + 1;
+            tx1[jss::SigningPubKey] = "";
+            tx1.removeMember(jss::Flags);
+            auto jt = env.jtnofill(
+                batch::outer(alice, seq, batchFee, tfAllOrNothing),
+                batch::inner_nofill(tx1),
```

Comment:
nit: I _think_ this will work?
```suggestion
            auto tx1 = pay(alice, bob, XRP(10));
            tx1[jss::Fee] = "0";
            tx1[jss::Sequence] = seq + 1;
            tx1[jss::SigningPubKey] = "";
            tx1.removeMember(jss::Flags);
            auto jt = env.jtnofill(
                batch::outer(alice, seq, batchFee, tfAllOrNothing),
                batch::inner(pay(alice, bob, XRP(10)), txflags(0)),
```

Reply:
(update: this doesn't work, going to see if I can figure out a way to make it work)

Reply:
thank you very much, updated: https://github.com/XRPLF/rippled/pull/5060/commits/230f1306f1c130fb221994d1a693dcddfe909e0c